docs/pjrt.md

```diff
@@ -100,6 +122,47 @@ for more information about TPU architecture.
   ([`gcloud compute tpus tpu-vm scp`](https://cloud.google.com/sdk/gcloud/reference/alpha/compute/tpus/tpu-vm/scp)) and run the code on each host in
   parallel (e.g. [`gcloud compute tpus tpu-vm ssh --workers=all
   --command="PJRT_DEVICE=TPU python run.py"`](https://cloud.google.com/sdk/gcloud/reference/alpha/compute/tpus/tpu-vm/ssh))
+- `xm.rendezvous` has been reimplemented using XLA-native collective
+  communication to enhance stability on large TPU pods. See below for more
+  details.
+
+### Changes to `xm.rendezvous`
+
+_New in PyTorch/XLA r1.14 (nightly only)_
+
+In practice, we found that running a single mesh master process was unreliable
+on TPU pods with thousands of chips due to the number of inbound connections to
+worker 0. A single client process timing out could cause a failure and force the
+entire workload to restart.
+
+Thus, we have reimplemented `xm.rendezvous` with native XLA collective
+communication, which is much more stable and well-tested on large TPU pods. This
+imposes two new constraints compared to the XRT implementation:
+
+- Because the payload has to become part of the XLA graph, `xm.mark_step` is
+  called both before and after the data is transferred. Calling `xm.rendezvous`
+  in the middle of model code may force an unwanted compilation.
+- Because XLA does not permit collective operations to run on a subset of
+  workers, all workers must participate in the `rendezvous`.
+
+If you require the old behavior of `xm.rendezvous` (i.e. communicating data
+without altering the XLA graph and/or synchronizing a subset of workers),
+consider using [`torch.distributed.barrier`](https://pytorch.org/docs/stable/distributed.html#torch.distributed.barrier)
+or [`torch.distributed.all_gather_object`](https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_gather_object)
+with a `gloo` process group. If you are also using the `xla` `torch.distributed`
+backend, you can use `torch.new_group` to create a `gloo` subgroup. See [this
+example](https://pytorch.org/docs/stable/distributed.html#torch.distributed.barrier)
+from the PyTorch documentation. Keep in mind these constraints:
+
+- `torch.distributed` is not fully supported on TPU v2/v3 at this time. Only a
+  subset of operations with the `xla` backend are tested, and `gloo` will likely
+  not work as expected in a multiprocessing context.
+- In our experiments, `gloo` does not scale well to thousands of chips, so
```

Comment:
Is there any documentation to read more about gloo with XLA? Just curious why gloo works.

Reply:
`gloo` works out of the box in this case because it doesn't interact with XLA at all. It's PyTorch's recommended backend for CPU-only training: https://pytorch.org/docs/stable/distributed.html#which-backend-to-use

Reply:
So you mean we recommend users to do cpu training or recommend users to move tensors to cpu and then use gloo?

Reply:
This section is talking specifically about `xm.rendezvous`, which is used for communicating data that's already on CPU. `gloo` won't work with XLA tensors directly, so you couldn't use it e.g. as the backend for DDP on TPUs. But you can still use `torch.distributed` utilities like `barrier` and `all_gather_object` in some cases as long as the input tensors aren't XLA tensors

Reply:
Oh, it's complicated but it makes sense.